### PR TITLE
chore: autoupdate finetuning

### LINKED ISSUE #3308: Autoupdate - missing update result on next launch (update un/succesful message, modal)
**Describe the bug**
User is not informed about the result of an update after relaunch.

**Info:**
 - Suite Version: any, app only (tested 21.1.0, commit ?)
 - OS: any (tested macOS 11.1 Big Sur)
 - Firmware Version: any (tested model T 2.3.5)

**Steps to reproduce:**
Precondition - update is available
1. Launch Suite and confirm update,
2. Select either update on next launch or update immediately,
3. Let either Suite relaunch or relaunch manually.

**Expected result:**
Suite displays an information about the result of the update (success, error) in form of a modal similar to how update is displayed, before entering passphrase etc.; would expect ~"update to version X succesful".

**Actual result:**
No info about update result displayed, either device connection prompt or wallet select modal is displayed.

**Screenshots:**
N/A since is doesn't exist

### PR DESCRIPTION
**Pull request by mroz22**
Commits: 
- abd3af182d65d262ed6546af34ea40520c4eeb38 finetuning of updating process as per @tsusanka request
- ~~3785b3433f0a851a36b08ef599bb36ba2b2cc624 this just downgrades app version so that abd3af182d65d262ed6546af34ea40520c4eeb38 can be easily tested with builds downloaded from CI as code changes in abd3af182d65d262ed6546af34ea40520c4eeb38 affect only PRE update behaviour of the app. This commit should be dropped eventually.~~ dropped
- 50cc06b7de6dbf6866a1268c2efe7f024f92a273 This should resolve #3308. I haven't tested it within full update scenario yet though. This will require doing some release draft and updating to it as the code changes affect POST update behaviour of the app.

### CODE REVIEW

File: packages/suite-desktop/src-electron/modules/auto-updater.ts [hunk 1]
```diff
@@ -127,7 +146,7 @@ const init = ({ mainWindow, store }: Dependencies) => {
         mainWindow.webContents.send('update/downloading', { ...progressObj });
     });
 
-    autoUpdater.on('update-downloaded', (info: any) => {
+    autoUpdater.on('update-downloaded', (info: UpdateDownloadedEvent) => {
```
Comment (goodhoko) [inline packages/suite-desktop/src-electron/modules/auto-updater.ts:163]: :+1:

File: packages/suite-desktop/src-electron/modules/auto-updater.ts [hunk 2]
```diff
@@ -44,13 +49,27 @@ const init = ({ mainWindow, store }: Dependencies) => {
     let isManualCheck = false;
     let latestVersion: Partial<UpdateInfo> = {};
     let updateCancellationToken: CancellationToken;
+    let shouldInstallUpdateOnQuit = false;
+
     const updateSettings = store.getUpdateSettings();
 
+    // update is downloaded on click and applied automatically whenever user restarts application (now or later)
     autoUpdater.autoDownload = false;
```
Comment (goodhoko) [inline packages/suite-desktop/src-electron/modules/auto-updater.ts]: The combination of the comment and the line beneath it is confusing for me. How does `autoDownload = false` affect `[is] applied automatically whenever user restarts application (now or later)`? Also, reading two lines further, it seems in direct contradiction with `autoUpdater.autoInstallOnAppQuit = false;`.

Please, correct me if I'm wrong, but maybe this comment fits better? (I used an imperative to denote that I'm describing an intention.)
```suggestion
    // Prevent downloading an update unless user explicitly asks for it.
    autoUpdater.autoDownload = false;
```

File: packages/suite-desktop/src-electron/modules/auto-updater.ts [hunk 3]
```diff
@@ -44,13 +49,27 @@ const init = ({ mainWindow, store }: Dependencies) => {
...
-    autoUpdater.autoInstallOnAppQuit = true;
+    // manually force install based on `shouldInstallUpdateOnQuit` var instead to have more control over the process
+    autoUpdater.autoInstallOnAppQuit = false;
```
Comment (goodhoko) [inline packages/suite-desktop/src-electron/modules/auto-updater.ts]: Why do we need `more control over the process`? Is it because the electron-updater is unreliable and we believe our handling will work better? Or something else? Either way, it would be cool to explain the motivation in the comment.

File: packages/suite-desktop/src-electron/modules/auto-updater.ts [hunk 4]
```diff
@@ -148,17 +167,33 @@ const init = ({ mainWindow, store }: Dependencies) => {
...
+                logger.info(
+                    'auto-updater',
+                    `Is configured to auto update after app quit ${shouldInstallUpdateOnQuit}`,
+                );
             });
     });
 
+    autoUpdater.on('before-quit-for-update', () => {
+        logger.info('auto-updater', 'before-quit-for-update event');
+    });
```
Comment (goodhoko) [inline packages/suite-desktop/src-electron/modules/auto-updater.ts:213]: Nitpick, optional: Personally, I'd omit this. If we want to get logs about what the `electron-updater` is doing then let's use its [dedicated api](https://github.com/trezor/trezor-suite/blob/b1e90d0712d37df9909dc380bceb19f7b1a259ac/packages/suite-desktop/src-electron/modules/auto-updater.ts#L52).

File: packages/suite-desktop/src-electron/modules/auto-updater.ts [hunk 5]
```diff
@@ -183,18 +218,16 @@ const init = ({ mainWindow, store }: Dependencies) => {
         autoUpdater
             .downloadUpdate(updateCancellationToken)
             .then(() => logger.info('auto-updater', 'Update downloaded'))
-            .catch(() => logger.info('auto-updater', 'Update cancelled'));
+            .catch(() => {
+                logger.info('auto-updater', 'Update cancelled');
+            });
     });
 
+    // todo: why do we need that? this is called somewhere from react? settings? is it still relevant?
```
Comment (goodhoko) [inline packages/suite-desktop/src-electron/modules/auto-updater.ts]: There's a [corresponding method](https://github.com/trezor/trezor-suite/blob/b1e90d0712d37df9909dc380bceb19f7b1a259ac/packages/suite-desktop/src-electron/preload.ts#L64) exposed within the React app and it's used at [two](https://github.com/trezor/trezor-suite/blob/b1e90d0712d37df9909dc380bceb19f7b1a259ac/packages/suite/src/views/settings/index.tsx#L112) [places](https://github.com/trezor/trezor-suite/blob/b1e90d0712d37df9909dc380bceb19f7b1a259ac/packages/suite-desktop/src/support/DesktopUpdater/Ready.tsx#L19) to *manually* trigger a restart

File: packages/suite-desktop/src-electron/modules/auto-updater.ts [hunk 6]
```diff
@@ -24,7 +24,6 @@ const preReleaseFlag = app.commandLine.hasSwitch('pre-release');
 
 const init = ({ mainWindow, store }: Dependencies) => {
     const { logger } = global;
-
```
Comment (goodhoko) [inline packages/suite-desktop/src-electron/modules/auto-updater.ts:22]: total nitpick: I'd prefer a smaller diff and keeping the blank line here.

File: packages/suite-desktop/src-electron/modules/auto-updater.ts [hunk 7]
```diff
@@ -223,7 +241,6 @@ const init = ({ mainWindow, store }: Dependencies) => {
             });
     });
 
-    // todo: why do we need that? this is called somewhere from react? settings? is it still relevant?
```
Comment (goodhoko) [inline packages/suite-desktop/src-electron/modules/auto-updater.ts]: This line was added in the previous commit and removed here. Can we not add it in the first place?

File: packages/suite-desktop/src-electron/modules/auto-updater.ts [hunk 8]
```diff
@@ -187,6 +199,12 @@ const init = ({ mainWindow, store }: Dependencies) => {
                     'auto-updater',
                     `Is configured to auto update after app quit ${shouldInstallUpdateOnQuit}`,
                 );
+
+                // save current app version so that after app is relaunch we can show info about transition to the new version
+                store.setUpdateSettings({
+                    ...store.getUpdateSettings(),
+                    preUpdateVersion: app.getVersion(),
+                });
```
Comment (goodhoko) [inline packages/suite-desktop/src-electron/modules/auto-updater.ts]: I'm wondering what stops us from doing this within the init step where we also dispatch the toast notification. It would require a bit less indirection and the logic would be confined to a single place.

The meaning of the `preUpdateVersion` variable would have to change slightly to something like `currentVersion` as it would be kept 1:1 with the actual version most of the time except the first start after an update.

In other words, I propose: let's have a `currentVersion` stored in the store. On every app start compare it with the app's actual version and if they differ dispatch a notification and update the store to the actual version.

What do you think?

**Review by goodhoko — CHANGES_REQUESTED** [review]
Generally Looks good. Below are some nitpicks as well as some bigger suggestions.

File: packages/suite-desktop/src-electron/modules/auto-updater.ts (hunk 5 above)
Comment (goodhoko) [reply packages/suite-desktop/src-electron/modules/auto-updater.ts]: From [this](https://github.com/trezor/trezor-suite/pull/3946#discussion_r663829881) I reckon you already found out. .) Resolving.

File: packages/suite-desktop/src-electron/modules/auto-updater.ts (hunk 6 above)
Comment (goodhoko) [reply packages/suite-desktop/src-electron/modules/auto-updater.ts:22]: feel free to RBY (resolve by yourself)

File: packages/suite-desktop/src-electron/modules/auto-updater.ts (hunk 4 above)
Comment (goodhoko) [reply packages/suite-desktop/src-electron/modules/auto-updater.ts:213]: RBY

File: packages/suite-desktop/src-electron/modules/auto-updater.ts (hunk 8 above)
Comment (mroz22) [reply packages/suite-desktop/src-electron/modules/auto-updater.ts]: I was considering this option as well but decided to avoid having variable `currentVersion` in store because it feels very confusing. Imagine you invetigate `store.ts` later and see `currentVersion` there. You would say: why the hell are we saving something we already know here? So thats why I decided to go with some name that clearly states its purpose and its value is set only during the update routine. That much for my motivation. 
Your solution has advantage of solving everything in one place. I don't know, should I rework it?

File: packages/suite-desktop/src-electron/modules/auto-updater.ts (hunk 3 above)
Comment (mroz22) [reply packages/suite-desktop/src-electron/modules/auto-updater.ts]: @tsusanka should I?

File: packages/suite-desktop/src-electron/modules/auto-updater.ts (hunk 8 above)
Comment (goodhoko) [reply packages/suite-desktop/src-electron/modules/auto-updater.ts]: Understand. Let's keep it as is. 👍
